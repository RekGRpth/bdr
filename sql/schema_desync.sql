-- RT#60660

SELECT * FROM public.bdr_regress_variables()
\gset

\c :writedb1

SELECT bdr.bdr_replicate_ddl_command($DDL$
CREATE TABLE public.desync (
   id integer primary key not null,
   n1 integer not null
);
$DDL$);

\d desync

SELECT bdr.wait_slot_confirm_lsn(NULL,NULL);

-- basic builtin datatypes
\c :writedb2

\d desync

\c :writedb1

-- Add a new (dropped) attribute on this node only
BEGIN;
SET LOCAL bdr.skip_ddl_replication = on;
SET LOCAL bdr.skip_ddl_locking = on;
ALTER TABLE desync ADD COLUMN dropme integer;
COMMIT;

BEGIN;
SET LOCAL bdr.skip_ddl_replication = on;
SET LOCAL bdr.skip_ddl_locking = on;
ALTER TABLE desync DROP COLUMN dropme;
COMMIT;

-- create natts=3 tuple on db1
--
-- This will fail to apply because natts of other side
-- is less than our natts.
INSERT INTO desync(id, n1) VALUES (1, 1);

SELECT * FROM desync ORDER BY id;

-- We cannot use a statement_timeout and wait for apply to prove
-- we're stuck here, because walsender connects/disconnects. But
-- a DDL lock attempt works.
-- This must ERROR not ROLLBACK
BEGIN;
SET LOCAL statement_timeout = '2s';
SELECT bdr.acquire_global_lock('ddl_lock');
ROLLBACK;

\c :writedb2

SELECT * FROM desync ORDER BY id;

-- create natts=2 tuple on db2
--
-- This should apply to writedb1 because we right-pad rows with low natts if
-- the other side col is dropped (or nullable)
INSERT INTO desync(id, n1) VALUES (2, 2);

SELECT bdr.wait_slot_confirm_lsn(NULL,NULL);

-- This must ROLLBACK not ERROR
BEGIN;
SET LOCAL statement_timeout = '10s';
SELECT bdr.acquire_global_lock('ddl_lock');
ROLLBACK;

SELECT * FROM desync ORDER BY id;

-- But now if we make matching changes on our side, our natts should match
-- and we can apply the remote's tuple too. For kicks this time do in one
-- txn.
BEGIN;
SET LOCAL bdr.skip_ddl_replication = on;
SET LOCAL bdr.skip_ddl_locking = on;
ALTER TABLE desync ADD COLUMN dropme integer;
ALTER TABLE desync DROP COLUMN dropme;
COMMIT;

\c :writedb1

-- So now this side should apply too
SELECT bdr.wait_slot_confirm_lsn(NULL,NULL);

-- This must ROLLBACK not ERROR
BEGIN;
SET LOCAL statement_timeout = '10s';
SELECT bdr.acquire_global_lock('ddl_lock');
ROLLBACK;

-- Yay!
SELECT * FROM desync ORDER BY id;

\c :writedb2
-- Yay! Again!
SELECT * FROM desync ORDER BY id;

SELECT bdr.bdr_replicate_ddl_command($DDL$
DROP TABLE public.desync;
$DDL$);
