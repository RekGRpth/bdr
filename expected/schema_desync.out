-- RT#60660
SELECT * FROM public.bdr_regress_variables()
\gset
\c :writedb1
SELECT bdr.bdr_replicate_ddl_command($DDL$
CREATE TABLE public.desync (
   id integer primary key not null,
   n1 integer not null
);
$DDL$);
 bdr_replicate_ddl_command 
---------------------------
 
(1 row)

\d desync
    Table "public.desync"
 Column |  Type   | Modifiers 
--------+---------+-----------
 id     | integer | not null
 n1     | integer | not null
Indexes:
    "desync_pkey" PRIMARY KEY, btree (id)

SELECT bdr.wait_slot_confirm_lsn(NULL,NULL);
 wait_slot_confirm_lsn 
-----------------------
 
(1 row)

-- basic builtin datatypes
\c :writedb2
\d desync
    Table "public.desync"
 Column |  Type   | Modifiers 
--------+---------+-----------
 id     | integer | not null
 n1     | integer | not null
Indexes:
    "desync_pkey" PRIMARY KEY, btree (id)

\c :writedb1
-- Add a new (dropped) attribute on this node only
BEGIN;
SET LOCAL bdr.skip_ddl_replication = on;
SET LOCAL bdr.skip_ddl_locking = on;
ALTER TABLE desync ADD COLUMN dropme integer;
COMMIT;
BEGIN;
SET LOCAL bdr.skip_ddl_replication = on;
SET LOCAL bdr.skip_ddl_locking = on;
ALTER TABLE desync DROP COLUMN dropme;
COMMIT;
-- create natts=3 tuple on db1
--
-- This will fail to apply because natts of other side
-- is less than our natts.
INSERT INTO desync(id, n1) VALUES (1, 1);
SELECT * FROM desync ORDER BY id;
 id | n1 
----+----
  1 |  1
(1 row)

-- We cannot use a statement_timeout and wait for apply to prove
-- we're stuck here, because walsender connects/disconnects. But
-- a DDL lock attempt works.
-- This must ERROR not ROLLBACK
BEGIN;
SET LOCAL statement_timeout = '2s';
SELECT bdr.acquire_global_lock('ddl_lock');
 acquire_global_lock 
---------------------
 
(1 row)

ROLLBACK;
\c :writedb2
SELECT * FROM desync ORDER BY id;
 id | n1 
----+----
  1 |  1
(1 row)

-- create natts=2 tuple on db2
--
-- This should apply to writedb1 because we right-pad rows with low natts if
-- the other side col is dropped (or nullable)
INSERT INTO desync(id, n1) VALUES (2, 2);
SELECT bdr.wait_slot_confirm_lsn(NULL,NULL);
 wait_slot_confirm_lsn 
-----------------------
 
(1 row)

-- This must ROLLBACK not ERROR
BEGIN;
SET LOCAL statement_timeout = '10s';
SELECT bdr.acquire_global_lock('ddl_lock');
 acquire_global_lock 
---------------------
 
(1 row)

ROLLBACK;
SELECT * FROM desync ORDER BY id;
 id | n1 
----+----
  1 |  1
  2 |  2
(2 rows)

-- But now if we make matching changes on our side, our natts should match
-- and we can apply the remote's tuple too. For kicks this time do in one
-- txn.
BEGIN;
SET LOCAL bdr.skip_ddl_replication = on;
SET LOCAL bdr.skip_ddl_locking = on;
ALTER TABLE desync ADD COLUMN dropme integer;
ALTER TABLE desync DROP COLUMN dropme;
COMMIT;
\c :writedb1
-- So now this side should apply too
SELECT bdr.wait_slot_confirm_lsn(NULL,NULL);
 wait_slot_confirm_lsn 
-----------------------
 
(1 row)

-- This must ROLLBACK not ERROR
BEGIN;
SET LOCAL statement_timeout = '10s';
SELECT bdr.acquire_global_lock('ddl_lock');
 acquire_global_lock 
---------------------
 
(1 row)

ROLLBACK;
-- Yay!
SELECT * FROM desync ORDER BY id;
 id | n1 
----+----
  1 |  1
  2 |  2
(2 rows)

\c :writedb2
-- Yay! Again!
SELECT * FROM desync ORDER BY id;
 id | n1 
----+----
  1 |  1
  2 |  2
(2 rows)

SELECT bdr.bdr_replicate_ddl_command($DDL$
DROP TABLE public.desync;
$DDL$);
 bdr_replicate_ddl_command 
---------------------------
 
(1 row)

